test/test_spark.py

```diff
@@ -31,70 +33,139 @@
+
+        for i in range(5):
+            assert(pytest.approx(scores[i], 0.01) == expected[i])
+
+    def test_job_cancelling(self):
+        from joblib import Parallel, delayed
+        import time
+        import tempfile
+        import os
+
+        tmp_dir = tempfile.mkdtemp()
+
+        def test_fn(x):
+            if x == 0:
+                # make the task-0 fail, then it will cause task 1/2/3 to be canceled.
+                raise RuntimeError()
+            else:
+                time.sleep(15)
+                # if the task finished successfully, it will write a flag file to tmp dir.
+                with open(os.path.join(tmp_dir, str(x)), 'w'):
+                    pass
+
+        with pytest.raises(Exception):
+            with parallel_backend('spark', n_jobs=2):
+                Parallel()(delayed(test_fn)(i) for i in range(2))
+
+        time.sleep(30)  # wait until we can ensure all task finish or cancelled.
+        # assert all jobs was cancelled, no flag file will be written to tmp dir.
+        assert len(os.listdir(tmp_dir)) == 0
+
+
+@unittest.skipIf(Version(pyspark.__version__).release < (3, 4, 0),
+                 "Resource group is only supported since spark 3.4.0")
+class TestGPUSparkCluster(unittest.TestCase):
+    @classmethod
+    def setup_class(cls):
+        cls.num_cpus_per_spark_task = 2
+        cls.num_gpus_per_spark_task = 1
+        gpu_discovery_script_path = os.path.join(
+            os.path.dirname(os.path.abspath(__file__)), "discover_2_gpu.sh"
+        )
+
+        cls.spark = (
+            SparkSession.builder.master("local-cluster[1, 2, 1024]")
+            .config("spark.task.cpus", "1")
+            .config("spark.task.resource.gpu.amount", "1")
+            .config("spark.executor.cores", "2")
+            .config("spark.worker.resource.gpu.amount", "2")
+            .config("spark.executor.resource.gpu.amount", "2")
+            .config("spark.task.maxFailures", "1")
+            .config(
+                "spark.worker.resource.gpu.discoveryScript", gpu_discovery_script_path
+            )
+            .getOrCreate()
+        )
+
+    @classmethod
+    def teardown_class(cls):
+        cls.spark.stop()
+
+    def test_resource_group(self):
+        def get_spark_context(x):
+            from pyspark import TaskContext
+            taskcontext = TaskContext.get()
+            assert taskcontext.cpus() == 1
+            assert len(taskcontext.resources().get("gpu").addresses) == 1
+            return TaskContext.get()
+
+        with parallel_backend('spark') as (ba, _):
+            Parallel(n_jobs=5)(delayed(get_spark_context)(i) for i in range(10))
+
+    def test_customized_resource_group(self):
+        def get_spark_context(x):
+            from pyspark import TaskContext
+            taskcontext = TaskContext.get()
+            assert taskcontext.cpus() == 2
+            assert len(taskcontext.resources().get("gpu").addresses) == 1
+            return TaskContext.get()
```

Comment:
Let's return other values (e.g. a number), `TaskContext` object might not be serializable and might cause issues in future spark versions.